pkg/apis/devops/v1alpha3/devopsproject_types.go

```diff
@@ -0,0 +1,73 @@
+/*
+Copyright 2019 The KubeSphere authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha3
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
+// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+
+const DevOpsProjectFinalizerName = "devopsproject.finalizers.kubesphere.io"
+
+const (
+	ResourceKindDevOpsProject     = "DevOpsProject"
+	ResourceSingularDevOpsProject = "devopsproject"
+	ResourcePluralDevOpsProject   = "devopsprojects"
+)
+
+// DevOpsProjectSpec defines the desired state of DevOpsProject
+type DevOpsProjectSpec struct {
```

Comment:
Yes, there is no spec yet, because it is an isolation concept similar to ns.
We need some specs (creator, etc.) similar to the workspace, I want to deal with that part after more discussion.

Reply:
DevOps project is a cluster-level resource.
Pipelines are namespace-level resources.
When we create a DevOps project, we will create an admin namespace for the project, and the pipeline will be stored in the admin namespace.